[DTensor] Assert DTensorSpec has valid placements

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158133

This helped identify buggy sharding rules during debugging, why not
check it in.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta @msaroufim @dcci

wconstab committed:
[DTensor] Assert DTensorSpec has valid placements

This helped identify buggy sharding rules during debugging, why not
check it in.

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Assert DTensorSpec has valid placements"


This helped identify buggy sharding rules during debugging, why not
check it in.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/_dtensor_spec.py:
```diff
@@ -32,6 +32,10 @@ class DTensorSpec:
             self.placements = tuple(self.placements)
+        if not len(self.placements) == self.mesh.ndim:
+            raise ValueError(
+                f"DTensorSpec requires one placement per mesh dim (ndim={self.mesh.ndim}), got {self.placements=}"
```
`mesh.ndim={self.mesh.ndim)` is better because `ndim` is ambiguous. It can be understood as `DTensorSpec.ndim` which is the tensor_meta ndim.

XilunWu reviewed (approved):
LGTM. Wonder if we have enforced this equation everywhere in DTensor.

wconstab committed:
Update on "[DTensor] Assert DTensorSpec has valid placements"


This helped identify buggy sharding rules during debugging, why not
check it in.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot rebase

wconstab committed:
Update on "[DTensor] Assert DTensorSpec has valid placements"


This helped identify buggy sharding rules during debugging, why not
check it in.

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta

[ghstack-poisoned]
